### PR TITLE
Add permanent-redirect and permanent-redirect-code annotations

### PR DESCRIPTION
**Pull request by Xinayder**
This PR adds 2 new annotations:
- `permanent-redirect`
- `permanent-redirect-code`

`permanent-redirect` will build a `static_response` handler for Caddy, redirecting requests to the value of the annotation field. 

`permanent-redirect-code` can be used to set custom HTTP codes for the redirect. It supports Caddy's values like `permanent` or `temporary`. If this annotation is not set, it defaults to 301.

#### Without `permanent-redirect-code`
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example
  annotations:
    kubernetes.io/ingress.class: caddy
    caddy.ingress.kubernetes.io/permanent-redirect: https://google.com
    caddy.ingress.kubernetes.io/backend-protocol: http
    caddy.ingress.kubernetes.io/disable-ssl-redirect: "true"
spec:
  rules:
  - host: example1.kubernetes.localhost
    http:
      paths:
      - path: /hello1
        pathType: Prefix
        backend:
          service:
            name: example1
            port:
              number: 8080
```

```log
$ curl -H 'Host: example1.kubernetes.localhost' http://127.0.0.1:8080/hello1 -v
*   Trying 127.0.0.1:8080...
* Connected to 127.0.0.1 (127.0.0.1) port 8080
> GET /hello1 HTTP/1.1
> Host: example1.kubernetes.localhost
> User-Agent: curl/8.4.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Alt-Svc: h3=":443"; ma=2592000
< Location: https://google.com
< Server: Caddy
< Date: Sat, 21 Oct 2023 14:58:19 GMT
< Content-Length: 0
< 
* Connection #0 to host 127.0.0.1 left intact
```

#### With `permanent-redirect-code` set to 308
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example
  annotations:
    kubernetes.io/ingress.class: caddy
    caddy.ingress.kubernetes.io/permanent-redirect: https://google.com
    caddy.ingress.kubernetes.io/permanent-redirect-code: "308"
    caddy.ingress.kubernetes.io/backend-protocol: http
    caddy.ingress.kubernetes.io/disable-ssl-redirect: "true"
spec:
  rules:
  - host: example1.kubernetes.localhost
    http:
      paths:
      - path: /hello1
        pathType: Prefix
        backend:
          service:
            name: example1
            port:
              number: 8080
```

```log
$ curl -H 'Host: example1.kubernetes.localhost' http://127.0.0.1:8080/hello1 -v
*   Trying 127.0.0.1:8080...
* Connected to 127.0.0.1 (127.0.0.1) port 8080
> GET /hello1 HTTP/1.1
> Host: example1.kubernetes.localhost
> User-Agent: curl/8.4.0
> Accept: */*
> 
< HTTP/1.1 308 Permanent Redirect
< Alt-Svc: h3=":443"; ma=2592000
< Location: https://google.com
< Server: Caddy
< Date: Sat, 21 Oct 2023 14:57:03 GMT
< Content-Length: 0
< 
* Connection #0 to host 127.0.0.1 left intact
```

### CODE REVIEW

**Review by mavimo — CHANGES_REQUESTED** [review]
Hey @Xinayder, thanks for your contribution!

I added some comments, let me know if you agree and eventually if I can help 😄

File: internal/caddy/ingress/redirect.go
[the remark this answers is not in the document]
Comment (Xinayder) [reply internal/caddy/ingress/redirect.go:1]: This is something I'd need help with. I couldn't find an example of testing annotations, but I think I can use the new `tls_test.go` as a base.

File: internal/caddy/ingress/redirect.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	ing := input.Ingress
+
+	var code string = "301"
+
+	// Logic taken from Caddy's parseRedir builtin function
+	redirectCode := getAnnotation(ing, permanentRedirectCode)
+	if redirectCode != "" {
+		switch redirectCode {
+		case "permanent":
+			code = "301"
```
[the remark this answers is not in the document]
Comment (Xinayder) [reply internal/caddy/ingress/redirect.go:34]: I just copied them from Caddy's code. I don't know the exact difference between both of them, but it seems like 308/307 are more strict, according to the RFC:

> Note: For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request. If this behavior is undesired, the [308 (Permanent Redirect)](https://www.rfc-editor.org/rfc/rfc9110.html#status.308) status code can be used instead.

File: internal/caddy/ingress/redirect.go
[the remark this answers is not in the document]
Comment (mavimo) [reply internal/caddy/ingress/redirect.go:1]: @Xinayder y, this is a good starting point, but LMK if you want me to commit an initial draft and then let you add the other tests.

[the remark this answers is not in the document]
Comment (Xinayder) [reply internal/caddy/ingress/redirect.go:1]: @mavimo That would help a lot. I'm stuck trying to figure out how to generate a Caddy config file from a dummy config passed through the test cases, but haven't had any success so far.

[the remark this answers is not in the document]
Comment (mavimo) [reply internal/caddy/ingress/redirect.go:1]: @Xinayder added some tests as example, let me know if is enough 😄

[the remark this answers is not in the document]
Comment (Xinayder) [reply internal/caddy/ingress/redirect.go:1]: @mavimo Thank you very much! I was to narrow sighted and didn't want to create a custom JSON file, my initial idea was to just generate a correct config, load it and check the resulting Caddy config object for the correct route/handler config, but your solution is simpler :smile:

File: internal/caddy/ingress/redirect.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xinayder) [reply internal/caddy/ingress/redirect.go:34]: @mavimo as I mentioned above, this is to keep compatibility with Caddy. If there's a preference for the stricter redirect codes, I'm fine with changing them, I don't have a preference.

File: internal/caddy/ingress/redirect_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			route, err := rp.IngressHandler(input)
+			if test.expectedError != "" {
+				if assert.Error(t, err) {
+					assert.EqualError(t, err, test.expectedError)
+				}
+			} else {
+				assert.NoError(t, err, "unable to generate the route.")
+			}
+
+			if test.expectedError == "" {
```
Comment (Xinayder) [inline internal/caddy/ingress/redirect_test.go:102]: I'm not confident with this way to check if we want the test to fail or not (having the `expectedError` field). If the field is present, we check if the error returned from the ingress plugin is what we expect.

If the field is not present, then we proceed with comparing the generated config and the sample config file if they match.

File: internal/caddy/ingress/redirect.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mavimo) [reply internal/caddy/ingress/redirect.go:34]: @mholt @Embraser01 do you have any strong opinion here?

File: internal/caddy/ingress/redirect_test.go (hunk 2 above)
Comment (mavimo) [reply internal/caddy/ingress/redirect_test.go:102]: Can I suggest to split this into two tests? One for the entry that do not expect any error and one for the group that expect errors? This think will make the tests more understandable and more clean in term of implementation, something like:
```
func TestRedirectConvertToCaddyConfig(t *testing.T) {
    // ...
}

func TestMisconfiguredRedirectConvertToCaddyConfig(t *testing.T) {
    // ...
}
```

File: internal/caddy/ingress/redirect.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		case "temporary":
+			code = "302"
+		default:
+			codeInt, err := strconv.Atoi(redirectCode)
+			if err != nil {
+				return nil, fmt.Errorf("not a supported redir code type or not valid integer: '%s'", redirectCode)
+			}
+
+			if codeInt < 300 || (codeInt > 399 && codeInt != 401) {
+				return nil, fmt.Errorf("redir code not in the 3xx range or 401: '%v'", codeInt)
```
Comment (mavimo) [inline internal/caddy/ingress/redirect.go:44]:
```suggestion
				return nil, fmt.Errorf("redirection code not in the 3xx range or 401: '%v'", codeInt)
```

**Comment by mavimo** [comment]
@Xinayder what do you think to change a bit approach here? I made a few research and having a `temporary` as redirect code for  a `permanent-redirect` annotation is a bit confusing to me. Can I propose to switch a bit the approach here and follow a more commont approach to have two different annotations for `permanent-redirect` and `temporary-redirect` as Nginx ingress controller does?
 * https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#permanent-redirect
 * https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#permanent-redirect-code
 * https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#temporal-redirect
I don't think is too much different from what you're doing here but make thing more understandable for a newcomer...

File: internal/caddy/ingress/redirect.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	var code string = "301"
+
+	// Logic taken from Caddy's parseRedir builtin function
+	redirectCode := getAnnotation(ing, permanentRedirectCode)
+	if redirectCode != "" {
+		switch redirectCode {
+		case "permanent":
+			code = "301"
+		case "temporary":
+			code = "302"
```
[the remark this answers is not in the document]
Comment (mavimo) [reply internal/caddy/ingress/redirect.go:36]: See https://github.com/caddyserver/ingress/pull/165#discussion_r1404820045

**Comment by mavimo** [comment]
PS: test are failing, I think tests do not reflect the code changes made in https://github.com/caddyserver/ingress/pull/165/commits/660815b7d5b30f2f3ce4720ac499c031c6162e29

**Comment by Xinayder** [comment]
@mavimo I was actually thinking of submitting a new PR including the temporal redirect but now I think it should be included in this one.

Regarding the behavior, how do we want to proceed? Keep Caddy compatibility or try to make it compatible with Nginx?

In Nginx they check if the custom code falls between the 300-308 range, if it doesn't, the code is set to the default, which is the 301 : https://github.com/kubernetes/ingress-nginx/blob/7f723c59855e82614582ff7b2efd1783b1afc2ee/internal/ingress/annotations/redirect/redirect.go#L129

**Comment by Xinayder** [comment]
I've added the `temporal-redirect` annotation to be compatible with Nginx. I don't mind changing it to `temporary-redirect`, but I guess it's nice to be compatible with nginx.

Other things I wanted to ask:
- Should I create custom error types for the plugin, so whenever their wording changes we don't need to change the tests?
- Should there be input validation for the annotation values, e.g., to check if they are valid URLs? Or is it already done by Caddy?

File: internal/caddy/ingress/redirect.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	if permanentRedir != "" {
+		// Logic taken from Caddy's parseRedir builtin function
+		redirectCode := getAnnotation(ing, permanentRedirectCodeAnnotation)
+		if redirectCode != "" {
+			switch redirectCode {
+			case "permanent":
+				code = "301"
+			case "temporary":
+				code = "302"
```
Comment (mavimo) [inline internal/caddy/ingress/redirect.go:46]: Do you think we still need it? As now we have specific annotation for permanent and temporary redirect, I suppose we can get rid of them.

**Comment by mavimo** [comment]
@Xinayder you're doing an amazing job!

IMHO we can avoid to create a custom error and to do domain validation (people may decide to use internal domain as redirect and is very hard to predict what is going to be allowed or not).

Eventually can be iterated in the future, WDYT?

File: internal/caddy/ingress/redirect.go (hunk 5 above)
Comment (Xinayder) [reply internal/caddy/ingress/redirect.go]: I've removed them now.

**Comment by Xinayder** [comment]
I think it's idiomatic to have custom error types in this case (correct me if I'm wrong). wrt validating URLs, you're right, validation would be a pain to setup to consider every possible case.

**Review by mavimo — APPROVED** [review]

**Pull request merged**